solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -1640,7 +1727,6 @@ public void unregisterCore(String collection) {
           if (v == null) return null;
           if (v.coreRefCount > 0) v.coreRefCount--;
           if (v.canBeRemoved()) {
-            watchedCollectionStates.remove(collection);
             lazyCollectionStates.put(collection, new LazyCollectionRef(collection));
             reconstructState.set(true);
             return null;
```

Comment:
So what happens when you return `null` in the compute? Does it delete the entry in the map or does it set the value to `null`?

If it's the later, then we do need the null checks in all of the methods that interact with the `collectionWatches` values.

Reply:
it removes the entry